### ISSUE TITLE
LogicException: The query result yielded cache contexts (user) that were not part of the static query analysis.

### ISSUE THREAD

**Issue by davidwhthomas** [drupal-graphql/graphql#569]
I'm getting this error when using the latest `8.x-3.0-beta6` and graphql_views `8.x-1.0-alpha4` modules

The error can be traced to:
_modules/contrib/graphql/src/GraphQL/Execution/QueryProcessor.php_ ~line 314

The output of 
```        
dpm($result->getCacheContexts());
dpm($metadata->getCacheContexts());
```
<img width="392" alt="screen shot 2018-04-04 at 2 50 24 pm" src="https://user-images.githubusercontent.com/1858243/38286117-b6dd4db2-3817-11e8-8a05-e6d8c8464cde.png">

It seems there's a stray `user` cache context added somewhere? I can't see anything related to the user being added in the simple view, so not sure but it's a fatal error;

**Comment by davidwhthomas** [comment]
While I'm not sure why `$metadata->getCacheContexts()` is different in the above, I've found a workaround as it was a fatal error.

Basically, the update excludes the problematic `user` cache context from the View while preserving the `user.permission` and `user.node_grants:view` cache contexts

Maybe there's another way to support the `user` cache context there, but for now, this patch is working from here.

Attached, with thanks.

[569.patch.txt](https://github.com/drupal-graphql/graphql/files/1874319/569.patch.txt)

**Comment by fubhy** [comment]
The problem is not the "stray" user context that is being added but the one that is MISSING from the pre-run-time cache contexts. ./cc @pmelab 

We build this as a safety-hatch to catch cases exactly like this one you discovered. With the new collection of cache metadata that runs on the query statically before the actual execution phase, we have a slight risk for cached results incorrectly leaking into the response. 

Basically, the error means that "The result contains a cache context that was not discovered before execution which means that we can't cache this response reliably without the risk of leaking potentially secure data to a wrong request context". Hence, the goal should be to fix the issue instead of the symptom :)

**Comment by bogue** [comment]
Yeah, running into a few of those too while porting to beta-6. I doubt it will be of any help -- this stuff being waaaay over my head -- but in my particular case, it affects "languages:language_url" context instead of "user" as in:
```
{ 
  menuByName(name: "main"){
    links{
      label
      url {
        path
      }
    }
  }
}
```
Removing the "url" reference makes it work (tough a little less useful ;). Sorry i can't be more helpful... Good luck!

**Comment by davidwhthomas** [comment]
@fubhy, thanks for the explanation. 

I understand and agree it would be best to fix the issue of the different cache contexts before/after execution there. The above patch to exclude the 'user' cache context is meant as just a temporary workaround for now. 

Let me know if some other suggestion there and I may be able to help with a patch.

Thanks also for the work on the excellent GraphQL module! it can a little painful to hear bug reports when there is so many nicely working features, overall - great stuff! cheers

**Comment by pmelab** [comment]
The cache information is emitted from the views execution result, so the views deriver misses something.

@davidwhthomas: Could you provide me with an example views configuration that produces this error?

**Comment by davidwhthomas** [comment]
Hi @pmelab thanks for checking, in general graphql_views is great!

Just this issue hoping to debug and fix.

I've attached a sample view export (`viewsNodeQuery`) 

Sample GraphQL query

```
query {
  viewsNodeQuery(pageSize: 1, filter: {nid: {value: "12"}, published: "1", type: "article"}) {
    results { entityId }
}
```

[569-graphql-views-viewsNodeQuery.yml.txt](https://github.com/drupal-graphql/graphql/files/1878813/569-graphql-views-viewsNodeQuery.yml.txt)

**Comment by pmelab** [comment]
Hi @davidwhthomas!

Thanks for the report! The problem is (now) very obvious. We just added cache information from the view configuration, but not from the executing display. All makes more sense now 😃 

I pushed a fix to the dev versions on GitHub and Drupal.org

https://github.com/drupal-graphql/graphql-views/commit/29378066117177ec963e9246e4caea7bfc222b85

**Comment by fubhy** [comment]
Thanks @pmelab! <3

**Comment by davidwhthomas** [comment]
Thanks @pmelab! I can confirm the issue is fixed 😄

**Comment by awm086** [comment]
I am getting similar error but with " The query result yielded cache contexts (url.query_args:key)", any I am using 8.x-3.0-beta6. The error is caused by tgmgt_content module. 
This occurs with anon users who have "execute arbitrary graphql queries" permission. 

$result->getCacheContexts() gives 
``` 
Array
(
    [0] => url.query_args:key
    [1] => user.node_grants:view
    [2] => user.permissions
)
```
And $metadata->getCacheContexts() gives: 
```
Array
(
    [0] => user.node_grants:view
    [1] => user.permissions
)

```

The error is caused by last line of access hook:

```
/**
 * Implements hook_entity_access().
 */
function tmgmt_content_entity_access(\Drupal\Core\Entity\EntityInterface $entity, $operation, \Drupal\Core\Session\AccountInterface $account) {
  $result = AccessResult::neutral();
  $key = \Drupal::request()->query->get('key');
  if ($entity instanceof ContentEntityInterface && $operation == 'view' && $key) {
    $entity = $entity->getUntranslated();
    $source_langcode = $entity->language()->getId();
    $current_job_items = tmgmt_job_item_load_latest('content', $entity->getEntityTypeId(), $entity->id(), $source_langcode);
    /** @var \Drupal\tmgmt\JobItemInterface $job_item */
    if ($current_job_items) {
      foreach ($current_job_items as $job_item) {
        $valid_key = \Drupal::service('tmgmt_content.key_access')
          ->getKey($job_item);
        if ($key === $valid_key && \Drupal::config('tmgmt.settings')->get('anonymous_access')) {
          $result = AccessResult::allowed();
        }
        $result->addCacheableDependency($job_item);
      }
      $result->addCacheableDependency(\Drupal::config('tmgmt.settings'));
    }
  }
  return  $result->addCacheContexts(['url.query_args:key']);
}
``` 

I wonder if this is an issue with graphql or tmgmt.
